src/Symfony/Component/Security/Core/Authorization/IsGrantedPayload.php

```diff
@@ -0,0 +1,39 @@
+<?php
+
+/*
+ * This file is part of the Symfony package.
+ *
+ * (c) Fabien Potencier <fabien@symfony.com>
+ *
+ * For the full copyright and license information, please view the LICENSE
+ * file that was distributed with this source code.
+ */
+
+namespace Symfony\Component\Security\Core\Authorization;
+
+use Symfony\Component\Security\Core\Authentication\AuthenticationTrustResolverInterface;
+use Symfony\Component\Security\Core\Authentication\Token\TokenInterface;
+use Symfony\Component\Security\Core\User\UserInterface;
+
+/**
+ * @template T
+ *
+ * @author Alexandre Daubois <alex.daubois@gmail.com>
+ */
+final class IsGrantedPayload
+{
+    public UserInterface $user;
+
+    /**
+     * @param T $subject
+     */
+    public function __construct(
+        public TokenInterface $token,
+        public mixed $subject,
+        public array $roleNames,
```

Comment:
```suggestion
        public array $attributes,
```
or is there a specific reason to only support roles?

Reply:
The idea was to match what's injected in the expression when one is given to IsGranted: https://symfony.com/doc/current/security/expressions.html

Reply:
Got it, thanks. @wouterj any thought on the preferred path?

Reply:
If we keep `$roleNames` instead of `$attributes`, we'll lead the developer to a RBAC, and leads him away from an ABAC. Maybe this is what we want, but in my opinion, the `CallableVoter` is a bit advanced and we can take advantage of this feature to move a bit to an ABAC system. WDYT?

Reply:
Indeed. Thinking more about this I think this should be about attributes, not roles.

Reply:
Indeed, let's not tight this to roles and make it consistent with how `isGranted` works instead.

Reply:
The reason a `role_names` variable is injected in expressions is because we were encouraging checking roles directly in the expression, at a time `is_granted()` was not available in them (at first). We should encourage using the authorizationChecker instead.

Adding `$attributes` does not make sense here. `role_names` in the ExpressionVoter is not about the attributes being checked (what is checked is the expression itself, and is the Closure in that case) but about the reacheable roles of the token. We should remove `$roles` without replacement